### PR TITLE
feat: allow skipping encryption and custom muxer factory in upgrader

### PR DESCRIPTION
**Pull request by ckousik**
Updates the `DefaultUpgrader` to use the new `UpgraderOptions` which allows skipping encryption and/or adding a custom muxer factory for transports which inherently support encryption and muxing.

### CODE REVIEW

**Comment by ckousik** [comment]
@mxinden @MarcoPolo

File: src/upgrader.ts
Comment (MarcoPolo) [inline src/upgrader.ts:280]: Can we make this a `const` declaration so we don't shadow outer vars (I think this is confusing).

File: src/upgrader.ts
Comment (MarcoPolo) [inline src/upgrader.ts:246]: Can we remove this declaration now that we don't need it here?

File: src/upgrader.ts
Comment (MarcoPolo) [inline src/upgrader.ts:302]: else if ?

**Review by MarcoPolo — CHANGES_REQUESTED** [review]
Can we also include the await for async newStream? https://github.com/libp2p/js-libp2p/pull/1396/files#diff-c199200bb38fc06cbe1654f84bb2cfaed1cf0aa9bc9ed71f646932604cbd9252R422

File: package.json
Comment (achingbrain) [inline package.json:81]:
```suggestion
    "lint": "aegir lint",
```

File: src/upgrader.ts
Comment (achingbrain) [inline src/upgrader.ts:273]: We still want to do connection protection here - it's use case is a little different to encrypting the whole channel, instead you're essentially making a private network on top of the encrypted connection because you'll need the PSK to communicate any data.

That is, I can support XYX connection encryption scheme (noise, etc) and connect to a remote node, but I also need to know the PSK to connect to a remote node that uses protected connections.

File: src/upgrader.ts
Comment (achingbrain) [inline src/upgrader.ts:337]: This value used to come from the connection encrypter's `encryptOutbound` return value.  Looks like now that's ignored in favour of the one extracted from the multiaddr.

This will be problematic because we want to support discovering the peer id during the noise handshake in future - please can you change it back to use the one from the connection encrypter if it's configured & returned.

**Review by achingbrain — CHANGES_REQUESTED** [review]

File: src/upgrader.ts
Comment (MarcoPolo) [reply src/upgrader.ts:273]: This will break if we're using native muxing since there isn't a single underlying "connection". But I agree that we should do this even with native encryption. Maybe skip this if we have a custom stream muxer? fwiw on go-libp2p the QUIC transport doesn't support PSK.

File: src/upgrader.ts
Comment (MarcoPolo) [inline src/upgrader.ts:294]: Maybe "native" instead?

File: src/upgrader.ts
Comment (ckousik) [reply src/upgrader.ts:273]: @achingbrain Please correct me if I have misunderstood here: For a connection that does not have native encryption, we use the `PreSharedKeyConnectionProtector` to wrap the plaintext stream. Then, the selected encryption scheme runs on top of the already encrypted stream. Is this done for nodes to limit which other nodes can connect to them?

As @MarcoPolo has mentioned, we do not have a single underlying stream in the webrtc spec. Instead, we create an ephemeral data channel for the Noise handshake. Subsequent streams only use the native DTLS encryption.

Comment (MarcoPolo) [reply src/upgrader.ts]: In go-libp2p, we don't support PSK in QUIC or WebTransport.

Comment (MarcoPolo) [reply src/upgrader.ts]: We probably can't support a PSK with native encryption without giving this more thought. I think this change is okay.

**Review by MarcoPolo — APPROVED** [review]

**Comment by MarcoPolo** [comment]
Can we move this out of draft @ckousik ?

File: package.json
Comment (achingbrain) [inline package.json:119]: These need to change to version numbers before merging

File: src/upgrader.ts
Comment (achingbrain) [inline src/upgrader.ts:266]: After the [suggested change](https://github.com/libp2p/js-libp2p-interfaces/pull/290#discussion_r988230911) in the interface repo we can make this decision on explicit settings rather than the lack of presence of an overriding muxer factory:

```suggestion
    if (opts.skipProtection !== true) {
```

File: src/upgrader.ts
Comment (achingbrain) [inline src/upgrader.ts:277]: I think this is a bit easier to read.  The linter will also probably complain about not handling boolean values explicitly.

```suggestion
      if (opts?.skipEncryption === false) {
```

File: src/upgrader.ts
Comment (achingbrain) [inline src/upgrader.ts:296]: Doe we need the extra parentheses?

```suggestion
      if (opts?.muxerFactory != null) {
```

File: src/upgrader.ts
Comment (achingbrain) [inline src/upgrader.ts:333]: I think this will upset the linter.  You can run `npm run lint` locally to check.

```suggestion
      remotePeer
```

**Review by achingbrain — CHANGES_REQUESTED** [review]
LGTM, some small nits and needs https://github.com/libp2p/js-libp2p-interfaces/pull/290 landing first.

File: src/upgrader.ts
Comment (achingbrain) [reply src/upgrader.ts]: See https://github.com/libp2p/js-libp2p-interfaces/pull/290#discussion_r988230911 - it's a small change but I think we need to split skipping connection protection out into it's own option - it makes the logic in the changes here easier to follow and also leaves the door open for transports that do their own encryption and muxing but also support PSK.

File: package.json
Comment (achingbrain) [inline package.json:118]:
```suggestion
    "@libp2p/interface-stream-muxer": "^3.0.0",
```

File: package.json
Comment (achingbrain) [inline package.json:119]:
```suggestion
    "@libp2p/interface-transport": "^2.0.0",
```

**Comment by achingbrain** [comment]
> I can't push to this branch.

Me neither - @ckousik I noticed the same thing about your [PR to `@libp2p/interfaces`](https://github.com/libp2p/js-libp2p-interfaces/pull/290) - could you please [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) on PRs in future?

**Comment by ckousik** [comment]
@achingbrain Should I close this PR and reopen ?

**Comment by achingbrain** [comment]
I think you should be able to grant permission by checking the box on the bottom right of this thread under the PR participants list.  I can't see the box but I see it on PRs I have open to other repos:

image

File: src/upgrader.ts
Comment (MarcoPolo) [reply src/upgrader.ts:277]: Wouldn't this return true if `opts` is undefined or `opts.skipEncryption` is undefined?

Comment (MarcoPolo) [reply src/upgrader.ts:277]: What I mean is that we want the default behavior to be to not skip encryption so we should do `opts?.skipEncryption !== true`. [example](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABFATjCBrAngCjgBygGcB+ALkQG8iMZ9yAjOOAGwFMBDMAXwEoqAUImEiYwRHkKkAdDTqIAvEsTAOLIm36UhI3RARFWbaSzgBzHACIAcnCiI5+fDDBnLvHbu6I26toN09AyMTcytHZ1d3TxFuATiBUEhYBEQGNigoNhRJYnIqR0Zmdi4+AMCxCQI82Vp8RABCZVQQTXLAxH0wQ3ZQixs7BzrItw8O718NdsCunuNTfoiXUZjhOITUdGwcD03MXEo+AQF0zOydk4ysnMPeIA)

**Comment by MarcoPolo** [comment]
I copied this to: https://github.com/libp2p/js-libp2p/pull/1411 with the requested changes.

**Comment by ckousik** [comment]
@achingbrain Since the fork is owned by the little-bear-labs org, I do not get that option at all.

**Comment by MarcoPolo** [comment]
superseded by #1411

**Pull request closed without merging**
